### PR TITLE
docs(frontend): add architecture docs

### PR DESCRIPTION
**Pull request by Bushuo**
Ref: #326 

This PR documents the frontend architecture in its current state. 

While writing this, I found some inconsistencies that we should fix:
- [ ] Remove the icons folder -> should be under assets
- [ ] Create a `lib` folder that holds the configurations for libraries (to follow bulletproof react - see [Frontend Architecture Pattern Decision](https://github.com/ElektraInitiative/PermaplanT/blob/master/doc/decisions/frontend_architecture_pattern.md))
- [ ] Merge `bindings` and `generated` folder.  -> 
  - I would prefer a structure like this:
  `generated/`
    - `bindings.ts`
    - `docs/`

@markus2330 we can either fix this here or in separate PRs. 
The frontend architecture is not as rigid as the backend's, so I kept it short. Should I add more detail?

## Basics

- [X] Details of what you changed are in commit messages.
- [X] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildserver is happy.
- [X] The PR is rebased with current master.

## Checklist

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I fixed all affected decisions
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I translated all strings visible to the user
- [ ] I mentioned [every code or binary](/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] Documentation is conforming to [our Documentation Guidelines](/doc/documentation.md)
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to our Coding Guidelines
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

### CODE REVIEW

File: doc/architecture/frontend.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# Frontend Architecture
+
+## Structure
+
+- `__mocks__/`: Contains mocked modules for tests
```
Comment (markus2330) [inline doc/architecture/frontend.md:5]: Please describe why it has the special name: Is this used by some framework or is this some convention?

File: doc/architecture/frontend.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+- `assets/`: Contains asset files like images
```
Comment (markus2330) [inline doc/architecture/frontend.md:6]: Describe that it only should contain svg or similar, larger images should be in Nextcloud.

File: doc/architecture/frontend.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+- `bindings/`: Contains the generated API bindings
```
Comment (markus2330) [inline doc/architecture/frontend.md:7]: Please describe: Who generates? How does it work?

File: doc/architecture/frontend.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+- `components/`: Contains components that are shared across features
```
Comment (markus2330) [inline doc/architecture/frontend.md:8]: Write that they need to be part of Storybook. Describe minimum requirements for documentation.

File: doc/architecture/frontend.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+- `config/`: Contains initialization code for the environment and libraries
+- `features/`: Contains the distinct features of the application
+  - `dark_mode/`: Contains the dark_mode feature (Switcher and store)
+  - `imprint_page/`: Contains the imprint page
+  - `landing_page/`: Contains the landing page
+  - `map_planning/`: Contains the map with layers
+  - `seeds/`: Contains the seed routes
```
Comment (markus2330) [inline doc/architecture/frontend.md:15]: I don't think we need this list.

File: doc/architecture/frontend.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
+# Frontend Architecture
+
+## Structure
+
+- `__mocks__/`: Contains mocked modules for tests
+- `assets/`: Contains asset files like images
+- `bindings/`: Contains the generated API bindings
+- `components/`: Contains components that are shared across features
+- `config/`: Contains initialization code for the environment and libraries
+- `features/`: Contains the distinct features of the application
```
Comment (markus2330) [inline doc/architecture/frontend.md:10]: Here some guideline what a "distinct features" is, and what the alternatives for implementations are.

File: doc/architecture/frontend.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  - `dark_mode/`: Contains the dark_mode feature (Switcher and store)
+  - `imprint_page/`: Contains the imprint page
+  - `landing_page/`: Contains the landing page
+  - `map_planning/`: Contains the map with layers
+  - `seeds/`: Contains the seed routes
+- `generated/`: Contains generated code, for now only docs
```
Comment (markus2330) [inline doc/architecture/frontend.md:16]: Why isn't bindings also a subfolder here? Here description of subfolders might make sense, as the list doesn't change quickly.

File: doc/architecture/frontend.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+- `hooks/`: Contains hooks that are shared across features
+- `icons/`: Contains svg icons
+- `routes/`: Contains the routes of the application
+- `styles/`: Contains the global and miscellaneous style sheets
+- `utils/`: Contains additional utility functions
```
Comment (markus2330) [inline doc/architecture/frontend.md:21]: Also here some guideline is needed. E.g. at least two components want to share the code or similar.

File: doc/architecture/frontend.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+  - `dark_mode/`: Contains the dark_mode feature (Switcher and store)
+  - `imprint_page/`: Contains the imprint page
+  - `landing_page/`: Contains the landing page
+  - `map_planning/`: Contains the map with layers
+  - `seeds/`: Contains the seed routes
+- `generated/`: Contains generated code, for now only docs
+- `hooks/`: Contains hooks that are shared across features
+- `icons/`: Contains svg icons
+- `routes/`: Contains the routes of the application
+- `styles/`: Contains the global and miscellaneous style sheets
```
Comment (markus2330) [inline doc/architecture/frontend.md:20]: Please describe: What is a "miscellaneous style sheet"?

File: doc/architecture/frontend.md [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  features -->|Uses hooks| hooks
+  features -->|Calls functions| utils
+  features -->|Uses icon files| icons
+  features -->|Uses asset files| assets
+  features -->|Uses some miscellaneous styles| styles
+  routes --> |Renders route features| features
+    subgraph f[example feature]
+      direction LR
+      E[index] -->|Components scoped to a specific feature| A[components]
+      E -->|API request declarations and api hooks related to a specific feature| B[api]
```
Comment (markus2330) [inline doc/architecture/frontend.md:38]: Please describe (probably already before): What exactly is `B[API]`? Where does it live?

```suggestion
      E -->|API request declarations and API hooks related to a specific feature| B[API]
```

File: doc/architecture/frontend.md [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+`__mocks__` and `generated` are not used during runtime
+
+### Features
+
+Most of the applications code lives here.
+Every folder created here represents a given feature and contains its domain specific code.
+Features are allowed to import another features public API which is exported from its `index.ts` file.
+Features are also allowed to import all other previously mentioned modules if needed.
+
+### Code documentation
```
Comment (markus2330) [inline doc/architecture/frontend.md:53]: See above, we need a description of how code is documented. Also write about `console.log` and assertions.

```suggestion
### Code Documentation
```

**Review by markus2330 — CHANGES_REQUESTED** [review]
Excellent job, I love it!

**Comment by markus2330** [comment]
> Remove the icons folder -> should be under assets

Great idea!

> Create a lib folder that holds the configurations for libraries (to follow bulletproof react - see [Frontend Architecture Pattern Decision](https://github.com/ElektraInitiative/PermaplanT/blob/master/doc/decisions/frontend_architecture_pattern.md))

Great idea, but there must be a clear distinction to utils etc. Imho it is not yet clear where to put what.

>    Merge bindings and generated folder. ->
>        I would prefer a structure like this:
>       generated/
>           bindings.ts
>           docs/

See comments, I had the same idea :wink: 

But then the sentence "`__mocks__` and `generated` are not used during runtime" also needs to be fixed, doesn't it?

> @markus2330 we can either fix this here or in separate PRs.

Better in a separate PR.

> The frontend architecture is not as rigid as the backend's, so I kept it short. Should I add more detail?

As short as possible but the basic questions must be answered. Especially it must be as clear as possible where to add which code and the general style of code (documentation, asserts, how generic, ...).

**Comment by Bushuo** [comment]
@markus2330 I included your suggestions in the latest commits. I added more details about the structure under the headings below the diagram.

> Great idea, but there must be a clear distinction to utils etc. Imho it is not yet clear where to put what.

I think it is clear from the context. If the method in question is configuration code for libraries it should be in `lib`. 
Like the axios configuration or the i18n configuration.
If it is a shared function that is not a hook it is a `util`.
If it is a shared hook then it should be in `hooks`.

> But then the sentence "__mocks__ and generated are not used during runtime" also needs to be fixed, doesn't it?

Yes I will fix this in the following PR.

**Review by markus2330 — APPROVED** [review]
Great work, thanks for the improvements!

**Comment by markus2330** [comment]
@buenaflor is this how you envisioned it (or better)?

**Comment by markus2330** [comment]
>> Great idea, but there must be a clear distinction to utils etc. Imho it is not yet clear where to put what.
>  I think it is clear from the context.

Imho only obvious parts of it.

> If the method in question is configuration code for libraries it should be in lib.
> Like the axios configuration or the i18n configuration.
> If it is a shared function that is not a hook it is a util.
> If it is a shared hook then it should be in hooks.

I think this text can also be included.

But there is more to it, the main question is: to which extend can/should we expose API from features and use them? If there is a heavy feature interaction the architecture quickly becomes a mess if we are not very careful. The guidelines should make clear what to add where so that we do not fall in such traps.

The architecture eventually should also include which features make use of APIs of other features.

Please create an issue describing what should be restructured or otherwise is not scope of this PR.

**Comment by Bushuo** [comment]
> Imho only obvious parts of it.

@markus2330 Ok, I need a little more time on this.
While researching I found a gem I would like to share that talks about architecture in a front-end (react) application.
https://martinfowler.com/articles/modularizing-react-apps.html

I think our current structure with Bulletproof React combines nicely with the principles discussed there.
The Payment Feature that is discussed in the article would be, in our case, one of several features.

> But there is more to it, the main question is: to which extend can/should we expose API from features and use them? If there is a heavy feature interaction the architecture quickly becomes a mess if we are not very careful. The guidelines should make clear what to add where so that we do not fall in such traps.

In my experience, the feature slices architecture tends to minimize interactions between features.
All concerns regarding one feature are self-contained, a case I can think of is, if the feature is service-like with multiple consumers, then there would be a public API and interactions.
In our case the dark-mode and map-store features are such service-like features, but I can not see the benefits of another architecture, like a strictly layered one, over ours.

> Please create an issue describing what should be restructured or otherwise is not scope of this PR.

**Comment by markus2330** [comment]
> While researching I found a gem I would like to share that talks about architecture in a front-end (react) application.
https://martinfowler.com/articles/modularizing-react-apps.html

Thank you for sharing! The main purpose of a software architect is to see such patterns even before they are already implemented and guide others to use them rightly.

> A case I can think of, where interactions would happen is, if the feature is service-like with multiple consumers.
In our case the dark-mode and map-store features are such service-like features, but I can not see the benefits of another architecture, like a strictly layered one, over ours.

The distinction of "service-like features" vs. features that do not share functionality is a very important one. Maybe we should even put dark-mode and map-store to "libs" instead to "features" to make this very clear? If not, it should be at least documented which features provide services to other features.

**Comment by markus2330** [comment]
> Yes, documenting this would be good. How about we document it in the index.ts files of the features?

If this means that it is also included in easily reachable parts of the Storybook it should be fine.

File: doc/architecture/frontend.md [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+- `__mocks__/`: Contains mocked modules for tests.
+- `assets/`: Contains small asset files like SVGs.
+- `bindings/`: Contains the generated API bindings.
+- `components/`: Contains components that are shared across features.
+- `config/`: Contains initialization code for the environment and libraries.
+- `features/`: Contains the distinct features of the application.
+- `generated/`: Contains generated code, for now only docs.
+- `hooks/`: Contains hooks that are shared across features.
+- `icons/`: Contains svg icons.
```
Comment (absurd-turtle) [inline doc/architecture/frontend.md:13]: I think the icons could be moved to assets

File: doc/architecture/frontend.md [hunk 13]
```diff
@@ -0,0 +1,87 @@
+# Frontend Architecture
+
+## Structure
+
+- `__mocks__/`: Contains mocked modules for tests.
+- `assets/`: Contains small asset files like SVGs.
+- `bindings/`: Contains the generated API bindings.
```
Comment (absurd-turtle) [inline doc/architecture/frontend.md:7]: Idk if bindings is the right word. Maybe dtos, shared_types or api_types would be more fitting

File: doc/architecture/frontend.md [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+### features
+
+Most of the application's code lives here.
+Every folder created here represents a given feature and contains its domain specific code.
+For example, if a feature interacts with the backend via network requests, it would have a sub module `api` that encapsulates this.
+
+Features are allowed to import another features public API which is exported from its `index.ts` file.
+Features are also allowed to import all other previously mentioned modules if needed.
+
+### Code Documentation
```
Comment (absurd-turtle) [inline doc/architecture/frontend.md:75]:
```suggestion
### code documentation
```

**Review by absurd-turtle — APPROVED** [review]
LGTM!

I have two notes about structure and wording.
Good job describing the structure though.

**Comment by absurd-turtle** [comment]
> While researching I found a gem I would like to share that talks about architecture in a front-end (react) application. https://martinfowler.com/articles/modularizing-react-apps.html

@Bushuo Did you add this somewhere in the docs? I think it would be nice to have it as a reference still after this PR is merged.

File: doc/architecture/frontend.md (hunk 12 above)
Comment (Bushuo) [reply doc/architecture/frontend.md:13]: Yes, I created a separate issue for this #355

File: doc/architecture/frontend.md (hunk 13 above)
Comment (Bushuo) [reply doc/architecture/frontend.md:7]: Yes, I think I will go with `api_types` because it is unambiguous, and is easy to understand.

**Comment by Bushuo** [comment]
> @Bushuo Did you add this somewhere in the docs? I think it would be nice to have it as a reference still after this PR is merged.

@absurd-turtle Not yet. Where do you think would be an appropriate place for it? I thought about the Helpful Links section. Maybe there is a better place?

File: doc/architecture/frontend.md (hunk 14 above)
Comment (markus2330) [reply doc/architecture/frontend.md:75]: @absurd-turtle the original capitalization is correct

File: doc/architecture/frontend.md (hunk 13 above)
Comment (markus2330) [reply doc/architecture/frontend.md:7]: I don't know `api_types` and would expect something like different types of APIs, like a functional one, one with locks, one lock-free etc. But this is probably not what should be there? If `api_types` is well-known in the React world I am okay with it, though. Please give a short explanation, though.

File: doc/architecture/frontend.md (hunk 14 above)
Comment (absurd-turtle) [reply doc/architecture/frontend.md:75]: I suggested it to be consistent with the other headlines but maybe the reason that they are not the same is that they represent the directories while this one does not.

File: doc/architecture/frontend.md [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+```
+
+`__mocks__` and `generated` are not used during runtime
+
+### \_\_mocks\_\_
+
+This folder has a special name to be recognized by our test runner jest.
+More details on how mocking with jest works can be found in the [jest documentation](https://jestjs.io/docs/manual-mocks).
+
+### assets
```
Comment (markus2330) [inline doc/architecture/frontend.md:47]:
```suggestion
### `assets`
```

File: doc/architecture/frontend.md (hunk 14 above)
Comment (markus2330) [reply doc/architecture/frontend.md:75]: I think the other headlines are folder names, and in such special situations (as it is case-sensitive) it is okay. Maybe they should be written like this (or see other comment):

```
##  `assets`
```

**Review by markus2330 — APPROVED** [review]

**Comment by markus2330** [comment]
jenkins build please

**Comment by Bushuo** [comment]
@markus2330 I think we can leave the `config` folder as is and not add a `lib` folder. As it has no real benefit over our current structure.
Library configurations are done in the `config` as before.

**Comment by markus2330** [comment]
jenkins build please

**Pull request merged**